File: 3427-special-array-ii/3427-special-array-ii.py
class Solution:
    def isArraySpecial(self, nums: List[int], queries: List[List[int]]) -> List[bool]:
        index = []
        n = len(nums)

        # Preprocess to find indices where parity is the same for adjacent elements
        prev = nums[0] % 2  # Check if the first element is odd
        for i in range(1, n):
            curr = nums[i] % 2  # Check if the current element is odd
            if prev == curr:    # Parity is the same
                index.append(i - 1)
            prev = curr

        # Process each query
        ans = []
        for query in queries:
            from_idx, to_idx = query

            # Binary search to find the first index >= from_idx
            idx = bisect_left(index, from_idx)
            if idx == len(index) or to_idx <= index[idx]:
                ans.append(True)  # No splits in the range
            else:
                ans.append(False)  # Split found in the range

        return ans

File: 3427-special-array-ii/3427-special-array-ii.cpp
class Solution {
public:
    vector<bool> isArraySpecial(vector<int>& nums, vector<vector<int>>& qs) {
        vector<int> index;
        int n = nums.size();

        bool prev = nums[0] & 1;   // odd

        for(int i=1; i<n; i++)
        {
            bool curr = nums[i] & 1;

            // parity is same
            if(prev == curr)
            {
                index.push_back(i-1);
            }

            prev = curr;
        }

        vector<bool> ans;
        for(auto q: qs)
        {
            int from = q[0], to = q[1];

            auto it = lower_bound(index.begin(), index.end(), from);
            if(it == index.end() || to <= *it)
            {
                ans.push_back(true);

            }
            else    
                ans.push_back(false);
        }

        return ans;
    }
};